tests/testing-framework/src/manifest_template.rs

```diff
@@ -0,0 +1,63 @@
+use anyhow::Context as _;
+use std::{path::Path, sync::OnceLock};
+
+use crate::TestEnvironment;
+
+/// A manifest template with template variables that can be substituted.
+pub struct ManifestTemplate {
+    manifest: String,
+}
+
+static TEMPLATE: OnceLock<regex::Regex> = OnceLock::new();
+impl ManifestTemplate {
+    /// Read a manifest template from a file.
+    pub fn from_file(path: impl AsRef<Path>) -> anyhow::Result<Self> {
+        let path = path.as_ref();
+        let manifest = std::fs::read_to_string(path)
+            .with_context(|| format!("could not read manifest template at '{}'", path.display()))?;
+        Ok(Self { manifest })
+    }
+
+    /// Substitute template variables in the manifest template.
+    pub fn substitute(&mut self, env: &mut TestEnvironment) -> Result<(), anyhow::Error> {
```

Comment:
If this ever gets more complicated, we might want to consider using an existing template library, e.g. https://crates.io/crates/tinytemplate (which Spin already depends on).